Add descriptions for new batch APIs (#33)
ably/sdk-api-reference: descriptions.md

||| `PaginatedResult` || A [`PaginatedResult`]{@link PaginatedResult} object containing an array of [`Stats`]{@link Stats} objects. |
| time() => io Time ||| RSC16 | Retrieves the time from the Ably service as milliseconds since the Unix epoch. Clients that do not have access to a sufficiently well maintained time source and wish to issue Ably [`TokenRequest`s]{@link TokenRequest} with a more accurate timestamp should use the [`queryTime`]{@link ClientOptions#queryTime} property instead of this method. |
||| `Time` || The time as milliseconds since the Unix epoch. |

> | batchPublish(BatchPublishSpec) => io BatchResult<BatchPublishSuccessResult \| BatchPublishFailureResult> ||| RSC22 | Publishes a [`BatchPublishSpec`]{@link BatchPublishSpec} object to one or more channels, up to a maximum of 100 channels. |

Author: idk if this change is for the best - in most of these rows there are no backticks (although there are some dotted around). it would be nice to be consistent but leaving them with no backticks is more aligned with the existing entries.

> | batchPublish(BatchPublishSpec) => io BatchResult<BatchPublishSuccessResult \| BatchPublishFailureResult> ||| RSC22 | Publishes a [`BatchPublishSpec`]{@link BatchPublishSpec} object to one or more channels, up to a maximum of 100 channels. |

Author: oh, i just realised that backticks are only used for generics types for some reason. not sure why we went with that but i'll update 🙂

| batchPublish([BatchPublishSpec]) => io [BatchResult<BatchPublishSuccessResult \| BatchPublishFailureResult>] ||| RSC22 | Publishes a [`BatchPublishSpec`]{@link BatchPublishSpec} object to one or more channels, up to a maximum of 100 channels. |
|| [`BatchPublishSpec`] ||| An array of [`BatchPublishSpec`]{@link BatchPublishSpec} objects. |
||| [`BatchResult<BatchPublishSuccessResult \| BatchPublishSuccessResult>`] || An array of [`BatchResult`]{@link BatchResult} objects containing information about the result of the batch publish for each requested channel for each provided [`BatchPublishSpec`]{@link BatchPublishSpec}. The array returned is in the same order as the provided [`BatchPublishSpec`]{@link BatchPublishSpec} array. |
| batchPresence([String]) => io BatchResult<BatchPresenceSuccessResult \| BatchPresenceFailureResult> ||| RSC23 | Retrieves the presence state for one or more channels, up to a maximum of 100 channels. Presence state includes the `clientId` of members and their current [`PresenceAction`]{@link PresenceAction}. |
|| [String] ||| An array of one or more channel names, up to a maximum of 100 channels. |

> ||| `BatchResult<BatchPresenceSuccessResult \| BatchPresenceFailureResult>` || A [`BatchResult`]{@link BatchResult} object containing information about the result of the batch presence request for each requested channel. |

Author: I think so, yes, although I'm not sure if this is a documented behaviour or not so it's probably best to recommend correlating requests to responses by matching on channel name. This is more important for token revocation and batch publishing since a `BatchPublishSpec[]` or `TokenRevocationTarget[]` could result in multiple success/failures for a single channel so matching on channel name wouldn't suffice.

> ||| `BatchResult<BatchPresenceSuccessResult \| BatchPresenceFailureResult>` || A [`BatchResult`]{@link BatchResult} object containing information about the result of the batch presence request for each requested channel. |

Reviewer: Ah ok, that makes sense!

| connect() ||| proxy for RTN11 | Calls [`connection.connect()`]{@link Connection#connect} and causes the connection to open, entering the connecting state. Explicitly calling `connect()` is needed if the [`autoConnect`]{@link ClientOptions#autoConnect} property is disabled. |
| time() => io Time ||| RTC6a | Retrieves the time from the Ably service as milliseconds since the Unix epoch. Clients that do not have access to a sufficiently well maintained time source and wish to issue Ably [`TokenRequest`s]{@link TokenRequest with a more accurate timestamp should use the [`queryTime`]{@link ClientOptions#queryTime} property instead of this method. |
||| `Time` || The time as milliseconds since the Unix epoch. |
| batchPublish(BatchPublishSpec) => io `BatchResult<BatchPublishSuccessResult \| BatchPublishFailureResult>` ||| RSC22 | Publishes a [`BatchPublishSpec`]{@link BatchPublishSpec} object to one or more channels, up to a maximum of 100 channels. |
|| `BatchPublishSpec` ||| A [`BatchPublishSpec`]{@link BatchPublishSpec} object. |

> ||| `BatchResult<BatchPublishSuccessResult \| BatchPublishFailureResult>` || A [`BatchResult`]{@link BatchResult} object containing information about the result of the batch publish for each requested channel. |

Reviewer 2: Should we explain how the return value of this method’s `results` property relates to the `BatchPublishSpec` that it received (in terms of ordering)?

> | batchPublish([BatchPublishSpec]) => io `[BatchResult<BatchPublishSuccessResult \| BatchPublishFailureResult>]` ||| RSC22 | Publishes a [`BatchPublishSpec`]{@link BatchPublishSpec} object to one or more channels, up to a maximum of 100 channels. |

Reviewer 2: ```suggestion
| batchPublish([BatchPublishSpec]) => io `[BatchResult<BatchPublishSuccessResult \| BatchPublishFailureResult>]` ||| RSC22 | Publishes one or more [`BatchPublishSpec`]{@link BatchPublishSpec} objects to one or more channels, up to a maximum of 100 channels. |
```

|| [`BatchPublishSpec`] ||| An array of [`BatchPublishSpec`]{@link BatchPublishSpec} objects. |
||| [`BatchResult<BatchPublishSuccessResult \| BatchPublishFailureResult>`] || An array of [`BatchResult`]{@link BatchResult} objects containing information about the result of the batch publish for each requested channel for each provided [`BatchPublishSpec`]{@link BatchPublishSpec}. The array returned is in the same order as the provided [`BatchPublishSpec`]{@link BatchPublishSpec} array. |

> | batchPresence([String]) => io `BatchResult<BatchPresenceSuccessResult \| BatchPresenceFailureResult>` ||| RSC23 | Retrieves the presence state for one or more channels, up to a maximum of 100 channels. Presence state includes the `clientId` of members and their current [`PresenceAction`]{@link PresenceAction}. |

Reviewer 2: Does this not return an array? That's what https://github.com/ably/specification/pull/139 shows.

Reviewer 2: Mistake in the spec, fixed by https://github.com/ably/specification/pull/162

> | batchPresence([String]) => io `BatchResult<BatchPresenceSuccessResult \| BatchPresenceFailureResult>` ||| RSC23 | Retrieves the presence state for one or more channels, up to a maximum of 100 channels. Presence state includes the `clientId` of members and their current [`PresenceAction`]{@link PresenceAction}. |

Reviewer 2: Should we explain how the return value of this method’s `results` property relates to the list of channel names that it received (in terms of ordering)?

Reviewer 2: I assume that the answer here is the same as the one for publish — the ordering is not guaranteed, you should match on the `channel` property instead.

|| `AuthOptions` ||| An [`AuthOptions`]{@link AuthOptions} object. |
||| `TokenDetails` || A [`TokenDetails`]{@link TokenDetails} object. |
|`TokenDetails`: TokenDetails?||| RSA16 | A [`TokenDetails`]{@link TokenDetails} object. |

> | revokeTokens(TokenRevocationTargetSpecifier, issuedBefore number?, allowReauthMargin boolean?) => io BatchResult<TokenRevocationSuccessResult \| TokenRevocationFailureResult>  ||| RSA17 | Revokes tokens as specified by the provided [`TokenRevocationTargetSpecifier`]{@link TokenRevocationTargetSpecifier}. Note, only tokens issued by an API key that had revocable tokens enabled before the token was issued. See the [token revocation docs](https://ably.com/docs/core-features/authentication#token-revocation) for more information. |

Reviewer: ```suggestion
| revokeTokens(TokenRevocationTargetSpecifier, issuedBefore number?, allowReauthMargin boolean?) => io `BatchResult<TokenRevocationSuccessResult \| TokenRevocationFailureResult>`  ||| RSA17 | Revokes the tokens specified by the provided [`TokenRevocationTargetSpecifier`]{@link TokenRevocationTargetSpecifier}. Only tokens issued by an API key that had revocable tokens enabled before the token was issued can be revoked. See the [token revocation docs](https://ably.com/docs/core-features/authentication#token-revocation) for more information. |
```

|| `TokenRevocationTargetSpecifier` ||| A [`TokenRevocationTargetSpecifier`]{@link TokenRevocationTargetSpecifier} object. |

> || `issuedBefore` ||| An optional number (Unix timestamp in milliseconds); if not specified it is set to the current time. The token revocation only applies to tokens issued before this timestamp. A request with an `issuedBefore` in the future, or more than an hour in the past will be rejected. |

Reviewer: ```suggestion
|| `issuedBefore` ||| A Unix timestamp in milliseconds where only tokens issued before this time are revoked. The default is the current time. Requests with an `issuedBefore` in the future, or more than an hour in the past, will be rejected. |
```

|| `allowReauthMargin` ||| If true, permits a token renewal cycle to take place without needing established connections to be dropped, by postponing enforcement to 30 seconds in the future, and sending any existing connections a hint to obtain (and upgrade the connection to use) a new token. The default is `false`, meaning that the effect is near-immediate. |

> ||| `BatchResult<TokenRevocationSuccessResult \| TokenRevocationFailureResult>` || A [`BatchResult`]{@link BatchResult} containing information about the result of the token revocation request.

Reviewer 2: It's not clear to me how to interpret this result — we pass a single specifier but receive a `BatchResult` with a `results` array. What do the contents of this array represent? We probably need to document that.

Reviewer 2: This single-specifier variant has since been removed from the spec.

> | revokeTokens([TokenRevocationTargetSpecifier], issuedBefore number?, allowReauthMargin boolean?) => io `BatchResult<TokenRevocationSuccessResult \| TokenRevocationFailureResult>`  ||| RSA17 | Revokes the tokens specified by the provided array of [`TokenRevocationTargetSpecifier`]{@link TokenRevocationTargetSpecifier}s. Only tokens issued by an API key that had revocable tokens enabled before the token was issued can be revoked. See the [token revocation docs](https://ably.com/docs/core-features/authentication#token-revocation) for more information. |

Reviewer 2: Does this not return an array? That's what https://github.com/ably/specification/pull/139 shows.

Reviewer 2: Mistake in the spec, fixed by https://github.com/ably/specification/pull/162

Contains information about the results of a batch operation.

| Method / Property | Parameter | Returns | Spec | Description |
|---|---|---|---|---|
| successCount: Int ||| BAR2a | The number of successful operations in the request. |
| failureCount: Int ||| BAR2b | The number of unsuccessful operations in the request. |
| results: [T] ||| BAR2c | An array of results for the batch operation. |

## class `BatchPublishSpec`

Describes the messages that should be published by a batch publish operation, and the channels to which they should be published.

| Method / Property | Parameter | Returns | Spec | Description |
|---|---|---|---|---|
| channels: [String] ||| BSP2a | The names of the channels to publish the `messages` to. |
| messages: [Message] ||| BSP2b | An array of [`Message`]{@link Message} objects. |

## class `BatchPublishSuccessResult`

Contains information about the result of successful publishes to a channel requested by a single [`BatchPublishSpec`]{@link BatchPublishSpec}.

| Method / Property | Parameter | Returns | Spec | Description |
|---|---|---|---|---|
| channel: String ||| BPR2a | The name of the channel the message(s) was published to. |

> | messageId: String ||| BPR2b | A unique ID prefixed to each published message. |

Reviewer 2: I think we need to say something about how to interpret this property, based on https://github.com/ably/specification/pull/139#discussion_r1184224904. Here's my badly-worded suggestion, perhaps @m-hulbert can help with the wording?

```suggestion
| messageId: String ||| BPR2b | An identifier shared by all of the messages published as a result of this `BatchPublishSpec` request. Clients receiving the message at index `i` of the `BatchPublishSpec.messages` array will see a `Message.id` property of `${messageId}.${i}`. |
```

Reviewer 2: I've updated it to "A unique ID prefixed to the [`id`]{@link Message#id} of each published message."

> | channel: String ||| BPR2a | The name of the channel |
> | messageId: String ||| BPR2b | The `messageId` prefix for the published message(s) |

Reviewer: ```suggestion
| channel: String ||| BPR2a | The name of the channel the message(s) was published to. |
| messageId: String ||| BPR2b | A unique ID prefixed to each published message. |
```

## class `BatchPublishFailureResult`

> Contains information about the result of unsuccessful publishes to a channel requested by a single `BatchPublishSpec`.

Reviewer: ```suggestion
Contains information about the result of unsuccessful publishes to a channel requested by a single [`BatchPublishSpec`]{@link BatchPublishSpec}.
```

| Method / Property | Parameter | Returns | Spec | Description |
|---|---|---|---|---|

> | channel: String ||| BPF2a | The name of the channel |

Reviewer: ```suggestion
| channel: String ||| BPF2a | The name of the channel the message(s) was published to. |
```

Author: This would be misleading since in this case the message(s) failed to publish, so i changed it to "the message(s) failed to publish to"

Reviewer: Good spot, thanks!

> | error: ErrorInfo ||| BPF2b | An `ErrorInfo` indicating the reason the message(s) failed to publish |

Reviewer: ```suggestion
| error: ErrorInfo ||| BPF2b | Describes the reason for which a message, or messages failed to publish to a channel as an [`ErrorInfo`]{@link ErrorInfo} object. |
```

## class `BatchPresenceSuccessResult`

Contains information about the result of a successful batch presence request for a single channel.

| Method / Property | Parameter | Returns | Spec | Description |
|---|---|---|---|---|

> | channel: String ||| BGR2a | The name of the channel |

Reviewer: ```suggestion
| channel: String ||| BGR2a | The channel name the presence state was retrieved for. |
```

> | presence: [PresenceMessage] ||| BGR2b | An array containing all members present on the channel |

Reviewer: ```suggestion
| presence: [PresenceMessage] ||| BGR2b | An array of [`PresenceMessage`]{@link PresenceMessage}s describing members present on the channel. |
```

## class `BatchPresenceFailureResult`

Contains information about the result of an unsuccessful batch presence request for a single channel.

| Method / Property | Parameter | Returns | Spec | Description |
|---|---|---|---|---|

> | channel: String ||| BGF2a | The name of the channel |

Reviewer: ```suggestion
| channel: String ||| BGF2a | The channel name the presence state was retrieved for. |
```

> | error: ErrorInfo ||| BGF2b | An `ErrorInfo` indicating the reason the presence request failed for the given channel |

Reviewer: ```suggestion
| error: ErrorInfo ||| BGF2b | Describes the reason for which presence state could not be retrieved for the channel as an [`ErrorInfo`]{@link ErrorInfo} object. |
```

## class `TokenRevocationTargetSpecifier`

Describes which tokens should be affected by a token revocation request.

| Method / Property | Parameter | Returns | Spec | Description |
|---|---|---|---|---|

> | type: String ||| TRT2a | The type of token revocation target specifier (eg. "clientId", "revocationKey", "channel") |

Reviewer: ```suggestion
| type: String ||| TRT2a | The type of token revocation target specifier. Valid values are `clientId`, `revocationKey` or `channel`. |
```

> | value: String ||| TRT2b | The value of the token revocation target specifier |

Reviewer: ```suggestion
| value: String ||| TRT2b | The value of the token revocation target specifier. |
```

## class `TokenRevocationSuccessResult`

Contains information about the result of a successful token revocation request for a single target specifier.

| Method / Property | Parameter | Returns | Spec | Description |
|---|---|---|---|---|

> | target: String ||| TRS2a | The target specifier |

Reviewer: ```suggestion
| target: String ||| TRS2a | The target specifier. |
```

| appliesAt: Int ||| TRS2b | A timestamp at which the token revocation will take effect |
| issuedBefore: Int ||| TRS2c | A timestamp for which tokens previously issued will be revoked |

## class `TokenRevocationFailureResult`

Contains information about the result of an unsuccessful token revocation request for a single target specifier.

| Method / Property | Parameter | Returns | Spec | Description |
|---|---|---|---|---|

> | target: String ||| TRF2a | The target specifier |

Reviewer: ```suggestion
| target: String ||| TRF2a | The target specifier. |
```

> | error: ErrorInfo ||| TRF2b | An `ErrorInfo` indicating the reason the token revocation request failed for the given specifier |

Reviewer: ```suggestion
| error: ErrorInfo ||| TRF2b | Describes the reason for which token revocation failed for the given `target` as an [`ErrorInfo`]{@link ErrorInfo} object. |
```
